File: go/indicators/welleswilder/directionalmovementminus/directionalmovementminus.go
package directionalmovementminus

import (
	"fmt"
	"math"
	"sync"

	"zpano/entities"
	"zpano/indicators/core"
)

const (
	dmmMnemonic    = "-dm"
	dmmDescription = "Directional Movement Minus"
)

// DirectionalMovementMinus is Welles Wilder's Directional Movement Minus indicator.
//
// The directional movement was developed in 1978 by Welles Wilder as an indication of trend strength.
//
// The calculation of the directional movement (+DM and −DM) is as follows:
//   - UpMove = today's high − yesterday's high
//   - DownMove = yesterday's low − today's low
//   - if DownMove > UpMove and DownMove > 0, then −DM = DownMove, else −DM = 0
//
// When the length is greater than 1, Wilder's smoothing method is applied:
//
//	Today's −DM(n) = Previous −DM(n) − Previous −DM(n)/n + Today's −DM(1)
//
// The indicator is not primed during the first length updates.
//
// Reference:
//
// Wilder, J. Welles Jr. (1978). New Concepts in Technical Trading Systems.
type DirectionalMovementMinus struct {
	mu           sync.RWMutex
	length       int
	noSmoothing  bool
	count        int
	previousHigh float64
	previousLow  float64
	value        float64
	accumulator  float64
	primed       bool
}

// NewDirectionalMovementMinus returns a new instance of the Directional Movement Minus indicator.
func NewDirectionalMovementMinus(length int) (*DirectionalMovementMinus, error) {
	if length < 1 {
		return nil, fmt.Errorf("invalid length %d: must be >= 1", length)
	}

	return &DirectionalMovementMinus{
		length:      length,
		noSmoothing: length == 1,
		value:       math.NaN(),
	}, nil
}

// Length returns the length parameter.
func (s *DirectionalMovementMinus) Length() int {
	return s.length
}

// IsPrimed indicates whether the indicator is primed.
func (s *DirectionalMovementMinus) IsPrimed() bool {
	s.mu.RLock()
	defer s.mu.RUnlock()

	return s.primed
}

// Metadata describes the output data of the indicator.
func (s *DirectionalMovementMinus) Metadata() core.Metadata {
	return core.BuildMetadata(
		core.DirectionalMovementMinus,
		dmmMnemonic,
		dmmDescription,
		[]core.OutputText{
			{Mnemonic: dmmMnemonic, Description: dmmDescription},
		},
	)
}

// Update updates the Directional Movement Minus given the next bar's high and low values.
func (s *DirectionalMovementMinus) Update(high, low float64) float64 {
	if math.IsNaN(high) || math.IsNaN(low) {
		return math.NaN()
	}

	if high < low {
		high, low = low, high
	}

	s.mu.Lock()
	defer s.mu.Unlock()

	if s.noSmoothing {
		if s.primed {
			deltaMinus := s.previousLow - low
			deltaPlus := high - s.previousHigh

			if deltaMinus > 0 && deltaPlus < deltaMinus {
				s.value = deltaMinus
			} else {
				s.value = 0
			}
		} else {
			if s.count > 0 {
				deltaMinus := s.previousLow - low
				deltaPlus := high - s.previousHigh

				if deltaMinus > 0 && deltaPlus < deltaMinus {
					s.value = deltaMinus
				} else {
					s.value = 0
				}

				s.primed = true
			}

			s.count++
		}
	} else {
		if s.primed {
			deltaMinus := s.previousLow - low
			deltaPlus := high - s.previousHigh

			if deltaMinus > 0 && deltaPlus < deltaMinus {
				s.accumulator += -s.accumulator/float64(s.length) + deltaMinus
			} else {
				s.accumulator += -s.accumulator / float64(s.length)
			}

			s.value = s.accumulator
		} else {
			if s.count > 0 && s.length >= s.count {
				deltaMinus := s.previousLow - low
				deltaPlus := high - s.previousHigh

				if s.length > s.count {
					if deltaMinus > 0 && deltaPlus < deltaMinus {
						s.accumulator += deltaMinus
					}
				} else {
					if deltaMinus > 0 && deltaPlus < deltaMinus {
						s.accumulator += -s.accumulator/float64(s.length) + deltaMinus
					} else {
						s.accumulator += -s.accumulator / float64(s.length)
					}

					s.value = s.accumulator
					s.primed = true
				}
			}

			s.count++
		}
	}

	s.previousLow = low
	s.previousHigh = high

	return s.value
}

// UpdateSample updates the Directional Movement Minus using a single sample value
// as a substitute for high and low.
func (s *DirectionalMovementMinus) UpdateSample(sample float64) float64 {
	return s.Update(sample, sample)
}

// UpdateScalar updates the indicator given the next scalar sample.
func (s *DirectionalMovementMinus) UpdateScalar(sample *entities.Scalar) core.Output {
	v := sample.Value

	output := make([]any, 1)
	output[0] = entities.Scalar{Time: sample.Time, Value: s.Update(v, v)}

	return output
}

// UpdateBar updates the indicator given the next bar sample.
func (s *DirectionalMovementMinus) UpdateBar(sample *entities.Bar) core.Output {
	output := make([]any, 1)
	output[0] = entities.Scalar{Time: sample.Time, Value: s.Update(sample.High, sample.Low)}

	return output
}

// UpdateQuote updates the indicator given the next quote sample.
func (s *DirectionalMovementMinus) UpdateQuote(sample *entities.Quote) core.Output {
	v := (sample.Bid + sample.Ask) / 2 //nolint:mnd

	output := make([]any, 1)
	output[0] = entities.Scalar{Time: sample.Time, Value: s.Update(v, v)}

	return output
}

// UpdateTrade updates the indicator given the next trade sample.
func (s *DirectionalMovementMinus) UpdateTrade(sample *entities.Trade) core.Output {
	v := sample.Price

	output := make([]any, 1)
	output[0] = entities.Scalar{Time: sample.Time, Value: s.Update(v, v)}

	return output
}

File: ts/indicators/welles-wilder/directional-movement-minus/directional-movement-minus.ts
import { Bar } from '../../../entities/bar';
import { Quote } from '../../../entities/quote';
import { Scalar } from '../../../entities/scalar';
import { Trade } from '../../../entities/trade';
import { Indicator } from '../../core/indicator';
import { IndicatorMetadata } from '../../core/indicator-metadata';
import { IndicatorOutput } from '../../core/indicator-output';
import { IndicatorType } from '../../core/indicator-type';
import { OutputMetadata } from '../../core/outputs/output-metadata';
import { OutputType } from '../../core/outputs/output-type';
import { DirectionalMovementMinusOutput } from './directional-movement-minus-output';

const dmmMnemonic = '-dm';
const dmmDescription = 'Directional Movement Minus';

/**
 * Welles Wilder's Directional Movement Minus indicator.
 *
 * The directional movement was developed in 1978 by Welles Wilder as an indication of trend strength.
 *
 * The calculation of the directional movement (+DM and −DM) is as follows:
 * - UpMove = today's high − yesterday's high
 * - DownMove = yesterday's low − today's low
 * - if DownMove > UpMove and DownMove > 0, then −DM = DownMove, else −DM = 0
 *
 * When the length is greater than 1, Wilder's smoothing method is applied:
 *   Today's −DM(n) = Previous −DM(n) − Previous −DM(n)/n + Today's −DM(1)
 *
 * The indicator is not primed during the first length updates.
 */
export class DirectionalMovementMinus implements Indicator {

  private readonly length_: number;
  private readonly noSmoothing: boolean;
  private count = 0;
  private previousHigh = 0;
  private previousLow = 0;
  private value_ = NaN;
  private accumulator = 0;
  private primed_ = false;

  constructor(length: number) {
    if (length < 1) {
      throw new Error(`invalid length ${length}: must be >= 1`);
    }

    this.length_ = length;
    this.noSmoothing = length === 1;
  }

  /** Returns the length parameter. */
  public get length(): number {
    return this.length_;
  }

  /** Indicates whether the indicator is primed. */
  public isPrimed(): boolean {
    return this.primed_;
  }

  /** Describes the output data of the indicator. */
  public metadata(): IndicatorMetadata {
    const outputMeta: OutputMetadata = {
      kind: DirectionalMovementMinusOutput.DirectionalMovementMinusValue,
      type: OutputType.Scalar,
      mnemonic: dmmMnemonic,
      description: dmmDescription,
    };

    return {
      type: IndicatorType.DirectionalMovementMinus,
      mnemonic: dmmMnemonic,
      description: dmmDescription,
      outputs: [outputMeta],
    };
  }

  /** Updates the Directional Movement Minus given the next bar's high and low values. */
  public update(high: number, low: number): number {
    if (isNaN(high) || isNaN(low)) {
      return NaN;
    }

    if (high < low) {
      const temp = high;
      high = low;
      low = temp;
    }

    if (this.noSmoothing) {
      if (this.primed_) {
        const deltaMinus = this.previousLow - low;
        const deltaPlus = high - this.previousHigh;

        if (deltaMinus > 0 && deltaPlus < deltaMinus) {
          this.value_ = deltaMinus;
        } else {
          this.value_ = 0;
        }
      } else {
        if (this.count > 0) {
          const deltaMinus = this.previousLow - low;
          const deltaPlus = high - this.previousHigh;

          if (deltaMinus > 0 && deltaPlus < deltaMinus) {
            this.value_ = deltaMinus;
          } else {
            this.value_ = 0;
          }

          this.primed_ = true;
        }

        this.count++;
      }
    } else {
      if (this.primed_) {
        const deltaMinus = this.previousLow - low;
        const deltaPlus = high - this.previousHigh;

        if (deltaMinus > 0 && deltaPlus < deltaMinus) {
          this.accumulator += -this.accumulator / this.length_ + deltaMinus;
        } else {
          this.accumulator += -this.accumulator / this.length_;
        }

        this.value_ = this.accumulator;
      } else {
        if (this.count > 0 && this.length_ >= this.count) {
          const deltaMinus = this.previousLow - low;
          const deltaPlus = high - this.previousHigh;

          if (this.length_ > this.count) {
            if (deltaMinus > 0 && deltaPlus < deltaMinus) {
              this.accumulator += deltaMinus;
            }
          } else {
            if (deltaMinus > 0 && deltaPlus < deltaMinus) {
              this.accumulator += -this.accumulator / this.length_ + deltaMinus;
            } else {
              this.accumulator += -this.accumulator / this.length_;
            }

            this.value_ = this.accumulator;
            this.primed_ = true;
          }
        }

        this.count++;
      }
    }

    this.previousLow = low;
    this.previousHigh = high;

    return this.value_;
  }

  /** Updates the Directional Movement Minus using a single sample value as a substitute for high and low. */
  public updateSample(sample: number): number {
    return this.update(sample, sample);
  }

  /** Updates the indicator given the next scalar sample. */
  public updateScalar(sample: Scalar): IndicatorOutput {
    const v = sample.value;
    const scalar = new Scalar();
    scalar.time = sample.time;
    scalar.value = this.update(v, v);
    return [scalar];
  }

  /** Updates the indicator given the next bar sample. */
  public updateBar(sample: Bar): IndicatorOutput {
    const scalar = new Scalar();
    scalar.time = sample.time;
    scalar.value = this.update(sample.high, sample.low);
    return [scalar];
  }

  /** Updates the indicator given the next quote sample. */
  public updateQuote(sample: Quote): IndicatorOutput {
    const v = (sample.bid + sample.ask) / 2;
    const scalar = new Scalar();
    scalar.time = sample.time;
    scalar.value = this.update(v, v);
    return [scalar];
  }

  /** Updates the indicator given the next trade sample. */
  public updateTrade(sample: Trade): IndicatorOutput {
    const v = sample.price;
    const scalar = new Scalar();
    scalar.time = sample.time;
    scalar.value = this.update(v, v);
    return [scalar];
  }
}
